x/dex/module.go

```diff
@@ -228,51 +230,25 @@ func (am AppModule) beginBlockForContract(ctx sdk.Context, contractAddr string)
 	ctx.Logger().Info(fmt.Sprintf("Orders %s, %s", am.keeper.Orders, contractAddr))
 
 	if isNewEpoch, currentEpoch := am.keeper.IsNewEpoch(ctx); isNewEpoch {
-		ctx.Logger().Info(fmt.Sprintf("Updating funding payment rate for epoch %d", currentEpoch))
-		for _, twap := range am.keeper.GetAllTwaps(ctx, contractAddr) {
-			dexPrice := twap.TwapPrice
-			ctx.Logger().Info(fmt.Sprintf("%s/%s: %d", twap.PriceDenom, twap.AssetDenom, dexPrice))
-			oraclePrice := uint64(100) // TODO: replace with oracle call
-			var diff uint64
-			var negative bool
-			if dexPrice < oraclePrice {
-				diff = oraclePrice - dexPrice
-				negative = true
-			} else {
-				diff = dexPrice - oraclePrice
-				negative = false
-			}
-			nativeSetFPRMsg := types.SudoSetFundingPaymentRateMsg{
-				SetFundingPaymentRate: types.SetFundingPaymentRate{
-					Epoch:      currentEpoch,
-					AssetDenom: twap.AssetDenom,
-					PriceDiff:  strconv.FormatUint(diff, 10),
-					Negative:   negative,
-				},
-			}
-			wasmMsg, err := json.Marshal(nativeSetFPRMsg)
-			if err != nil {
-				ctx.Logger().Info(err.Error())
+		ctx.Logger().Info(fmt.Sprintf("Updating price for epoch %d", currentEpoch))
+		priceRetention := am.keeper.GetParams(ctx).PriceSnapshotRetention
+		for _, pair := range am.keeper.GetAllRegisteredPairs(ctx, contractAddr) {
+			lastEpochPrice, exists := am.keeper.GetPriceState(ctx, contractAddr, currentEpoch-1, pair)
+			if exists {
+				newEpochPrice := types.Price{
+					SnapshotTimestampInSeconds: uint64(ctx.BlockTime().Unix()),
+					Pair:                       &pair,
+					Price:                      lastEpochPrice.Price,
+				}
+				am.keeper.SetPriceState(ctx, newEpochPrice, contractAddr, currentEpoch)
 			}
 
-			ctx.Logger().Info("Setting funding payment rate")
-			am.callClearingHouseContractSudo(ctx, wasmMsg, contractAddr)
-
-			var newPrices []uint64
-			if len(twap.Prices) == 24 { // replace with config
-				newPrices = append(twap.Prices[1:], twap.Prices[len(twap.Prices)-1])
-			} else {
-				newPrices = append(twap.Prices, twap.Prices[len(twap.Prices)-1])
+			// condition to prevent unsigned integer overflow
+			if currentEpoch >= priceRetention {
```

Comment:
is it always guaranteed that an epoch length is 60s? wouldnt it be good to add a check here that we only delete the epoch data if the current BlockTime - epoch BlockTime > 60 min? however one issue with that is that there is no mechanism to go back to an epoch if you skip over it on that condition :/ WDYT @codchen